scwidgets/widgets.py

```diff
@@ -0,0 +1,230 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+# Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+# @file
+# @author Matthew Andrew
+
+import ipywidgets as widgets
+import scipp as sc
+from scipp.plot import plot
+from .converters import scipp_object
+from IPython.core.display import display, HTML
+from typing import (Any, Mapping, Callable, Sequence, MutableMapping)
+
+InputType = Mapping[str, Callable[[str], Any]]
+
+
+class ProcessWidget(widgets.Box):
+    """
+    Provides a simple graphical wrapper around a given callable.
+    """
+    def __init__(self,
+                 scope: MutableMapping[str, Any],
+                 callable: Callable,
+                 name: str,
+                 inputs: InputType,
+                 descriptions: Mapping[str, str] = {},
+                 options: Mapping[str, Sequence[str]] = {}):
+        """
+        Parameters:
+        scope (dict): The scope you wish to add the outputs of the function to
+        callable (Callable): The function to call
+        name: name of widget to display
+        inputs (dict): dict of parameter_name: paramter_converter
+        descriptions (dict): dict of parameter_name: paramter_description.
+        If a parameter_name does not appear in the dict if will be used as
+        it's own description
+        options (dict): a dict of parameter_name: parameter_options if you wish
+        to display a list of options for a given parameter.
+        """
+        super().__init__()
+        self.scope = scope
+        self.callable = callable
+        self.inputs = inputs
+        self.input_widgets = []
+        self._setup_input_widgets(inputs, descriptions, options)
+
+        self.output = widgets.Text(placeholder='output name',
+                                   value='',
+                                   continuous_update=False)
+
+        self.button = widgets.Button(description=name)
+        self.button.on_click(self._on_button_clicked)
+
+        self.output_area = widgets.Output()
+
+        self.children = [
+            widgets.VBox([
+                widgets.HBox(self.input_widgets + [self.output, self.button]),
+                self.output_area
+            ])
+        ]
+
+        self.subscribers = []
```

Comment:
Should we consider removing the subscriber mechanism? This seems like a can of worms that I feel I would like to avoid for the first iteration?

Reply:
Agreed.